### PR TITLE
feat(status-api): Overall Status endpoint 

### PR DESCRIPTION
**Pull request by siradji**
Signed-off-by: Suraj Auwal <siradjiawoual@gmail.com>

#### Which issue(s) does this PR fixes?:

Fixes https://github.com/DeFiCh/jellyfish/issues/1274

### CODE REVIEW

File: apps/status-api/src/controllers/OverallStatusController.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  @Get()
+  async getAggregateStatus (): Promise<{ status: BlockchainStatus }> {
+    const liveness = await this.probe.liveness()
+
+    if (liveness.whale.status === 'down') {
+      return {
+        status: 'outage'
+      }
+    }
```
Comment (nichellekoh) [inline apps/status-api/src/controllers/OverallStatusController.ts:32]: On top of the liveness, can we also query readiness to check for the status?

File: apps/status-api/README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # DeFiChain Status API
 
-DeFiChain Status API, providing the statuses of different DeFiChain services
+DeFiChain Status API, providing the statuses of different DeFiChain services.
```
Comment (jingyi2811) [inline apps/status-api/README.md:3]: The changes in this file have both statuses and status. 
I think it should be either all statuses or all status.
I may miss the context, it is your call anyway to change this.

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+async function getWhaleStatus (status: HealthIndicatorStatus): Promise<HealthIndicatorResult> {
+  return {
+    whale: {
+      status
+    }
+  }
+}
+
+async function getBlockResponseWithPresetTime (minutesDiff: number): Promise<ApiPagedResponse<any>> {
```
Comment (jingyi2811) [inline apps/status-api/__test__/controllers/OverallStatusController.test.ts:93]: Typescript hates any
```suggestion
async function getBlockResponseWithPresetTime (minutesDiff: number): Promise<ApiPagedResponse<Block>> {
```
Block should get from 
`import { Block } from '@defichain/whale-api-client/dist/api/Blocks'`

Apply to other same code in the project

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  const blockTime = Date.now() / 1000 - (minutesDiff * 60)
```
Comment (jingyi2811) [inline apps/status-api/__test__/controllers/OverallStatusController.test.ts:94]: It is good to throw an error if (minutesDiff * 60) > Date.now() / 1000. This is just a suggestion. What do you think?

File: apps/status-api/src/controllers/OverallStatusController.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+   *  'degraded' when blockchain service is degraded.
+   *
+   *  @return {Promise<{ status: BlockchainStatus }>}
+   */
+
+  @Get()
+  async getAggregateStatus (): Promise<{ status: BlockchainStatus }> {
+    const liveness = await this.probe.liveness()
+
+    if (liveness.whale.status === 'down') {
```
Comment (jingyi2811) [inline apps/status-api/src/controllers/OverallStatusController.ts:28]:
```suggestion
    if (liveness?.whale?.status === 'down') {
```

Comment (siradji) [reply apps/status-api/src/controllers/OverallStatusController.ts:28]: I think that is a bit redundant. `WhaleApiIndicator` will always return `HealthIndicatorResult`

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts (hunk 3 above)
Comment (siradji) [reply apps/status-api/__test__/controllers/OverallStatusController.test.ts:93]: Nice catch there.

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts (hunk 4 above)
Comment (siradji) [reply apps/status-api/__test__/controllers/OverallStatusController.test.ts:94]: That would be great, but `minutesDiff` is a relatively small number in the testing scenario so throwing an error will be a bit redundant.

File: apps/status-api/src/controllers/OverallStatusController.ts (hunk 1 above)
Comment (siradji) [reply apps/status-api/src/controllers/OverallStatusController.ts:32]: Done!

File: apps/status-api/README.md (hunk 2 above)
Comment (siradji) [reply apps/status-api/README.md:3]: Yeah, it is a misunderstanding. `Defichain Status API` provides the statuses of Ocean and blockchain services.

**Review by jingyi2811 — DISMISSED** [review]

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  it('/overall - should get operational when Ocean and blockchain are up', async () => {
+    jest
+      .spyOn(apiTesting.app.get(WhaleApiProbeIndicator), 'liveness')
+      .mockReturnValueOnce(getWhaleStatus('up'))
+    jest
+      .spyOn(apiTesting.app.get(WhaleApiProbeIndicator), 'readiness')
+      .mockReturnValueOnce(getWhaleStatus('up'))
+    jest
+      .spyOn(apiTesting.app.get(WhaleApiClient).blocks, 'list')
+      .mockReturnValueOnce(getBlockResponseWithPresetTime(25))
```
Comment (jingyi2811) [inline apps/status-api/__test__/controllers/OverallStatusController.test.ts:26]: I don't understand these magic numbers: 25, 36, and 46. Mind to enlighten?

Comment (joeldavidw) [reply apps/status-api/__test__/controllers/OverallStatusController.test.ts:26]: The status of the blockchain is determined according to this. 
If a block is mined:
- `< 30 mins`: operational
- `30-45 mins`: degraded
- `> 45 mins`: outage

**Review by joeldavidw — APPROVED** [review]

**Review by eli-lim — DISMISSED** [review]

File: apps/status-api/__test__/controllers/OverallStatusController.test.ts (hunk 6 above)
Comment (cwkang1998) [reply apps/status-api/__test__/controllers/OverallStatusController.test.ts:26]: In this case I think its better to add some comments to give more context to the tests.

File: apps/status-api/src/controllers/OverallStatusController.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      }
+    }
+
+    const readiness = await this.probe.readiness()
+
+    if (readiness.whale.status === 'down') {
+      return {
+        status: 'outage'
+      }
+    }
```
Comment (cwkang1998) [inline apps/status-api/src/controllers/OverallStatusController.ts:40]: Just some refactoring
```suggestion
    const liveness = await this.probe.liveness()
    const readiness = await this.probe.readiness()
    
    if (liveness.whale.status === 'down' || readiness.whale.status === 'down') {
      return {
        status: 'outage'
      }
    }
```

Comment (jingyi2811) [reply apps/status-api/src/controllers/OverallStatusController.ts:40]: Maybe it is better not to use the same status

**Comment by siradji** [comment]
@eli-lim , added comments for better clarity.

**Review by eli-lim — APPROVED** [review]

File: apps/status-api/src/controllers/OverallStatusController.ts (hunk 7 above)
Comment (eli-lim) [reply apps/status-api/src/controllers/OverallStatusController.ts:40]: boolean short circuit can prevent the second redundant call
```
if (
  (await this.probe.liveness()).whale.status === 'down' || 
  (await this.probe.readiness()).whale.status === 'down'
) {
  return {
    status: 'outage'
  }
}
```

Albeit at the expense of some readability

**Review by cwkang1998 — APPROVED** [review]

**Review by joeldavidw — APPROVED** [review]

**Review by eli-lim — APPROVED** [review]

**Review by cwkang1998 — APPROVED** [review]

File: apps/status-api/src/controllers/OverallStatusController.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   *  To provide overall status for Ocean and blockchain services. Returns
+   *  'operational' when both services are up,
+   *  'outage' when either service is down and
+   *  'degraded' when blockchain service is degraded.
+   *
+   *  @return {Promise<{ status: BlockchainStatus }>}
+   */
+
```
Comment (fuxingloh) [inline apps/status-api/src/controllers/OverallStatusController.ts:23]:
```suggestion
```

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**